src/trait_group.h

```diff
@@ -0,0 +1,202 @@
+ */
+Trait_list traits_from( const Trait_group_tag &gid );
+
+/**
+ * Check whether a particular trait is a member of the given group.
+ */
+bool group_contains_trait( const Trait_group_tag &gid, const trait_id &tid );
+
+/**
+ * Check whether the given group ID has been defined.
+ */
+bool group_is_defined( const Trait_group_tag &gid );
+
+/**
+ * See @ref mutation_branch::load_trait_group
+ */
+void load_trait_group( JsonObject &jsobj, const Trait_group_tag &gid, const std::string &subtype );
+
+
+/**
+ * Get a trait group ID and optionally load an inlined trait group.
+ *
+ * If the next value in the JSON stream is string, it's assumed to be a trait group id and it's
+ * returned directly.
+ *
+ * If the next value is a JSON object, it is loaded as a trait group. The group will be given a
+ * unique id (if the JSON object contains an id, it is ignored) and that id will be returned.
+ * If the JSON object does not contain a subtype, the given default is used.
+ *
+ * If the next value is a JSON array, it is loaded as a trait group: the default_subtype will be
+ * used as subtype of the new trait group and the array is loaded like the "entries" array of
+ * a trait group definition (see format of trait groups).
+ *
+ * @param stream Stream to load from
+ * @param default_subtype If an inlined trait group is loaded this is used as the default
+ * subtype. It must be either "distribution" or "collection". See @ref Trait_group.
+ * @throw std::string as usual for JSON errors, including invalid input values.
+ */
+Trait_group_tag load_trait_group( JsonIn &stream, const std::string &default_subtype );
+
+/**
+ * Show a debug menu for testing trait groups.
+ */
+void debug_spawn();
+
+}
+
+/**
+ * Base interface for trait generation.
+ * Used to generate a list of traits.
+ */
+class Trait_creation_data
+{
+    public:
+        typedef std::vector<Trait_group_tag> RecursionList;
+
+        Trait_creation_data( int _probability ) : probability( _probability ) {}
+        virtual ~Trait_creation_data() {};
+
+        /**
+         * Create a list of traits. The resulting list might be empty.
+         * @param[out] rec Recursion list, output goes here
+         */
+        virtual Trait_list create( RecursionList &rec ) const = 0;
+        Trait_list create() const;
+
+        /**
+         * Check the trait templates for consistency (valid trait types, etc).
+         */
+        virtual void check_consistency() const = 0;
+
+        /**
+         * Remove the given trait from this and all linked groups.
+         * Used for blacklisting.
+         */
+        virtual bool remove_trait( const trait_id &tid ) = 0;
+
+        /**
+         * Check if this group or any linked groups contain the given trait.
+         */
+        virtual bool has_trait( const trait_id &tid ) const = 0;
+
+        // Probability, used by the parent object.
+        int probability;
+    private:
+        // not implemented
```

Comment:
Than just delete them (add ` = delete;`) instead of making them private. The code in "item_group.h" is old, so it doesn't use features from C++11, new code like this should use them where appropriate.